### DISCUSSION TITLE
v4.0.0

### DISCUSSION THREAD

**Discussion by coolapso** [streamdeck-linux-gui/streamdeck-linux-gui#115] [Announcements]
# Hi everyone,

Version 4.0.0 is out, and it is packed with new features and bug fixes!

> [!WARNING]  
> Please be aware that this is a major version release and contains breaking changes that are not backwards compatible. The application will attempt to create a backup of your configuration, but please ensure you also make one yourself.

There are significant changes in this release, with the most important feature being the ability to have buttons with multiple states! Additionally, we have added support for default configuration by device type.

You can find the release notes [here](https://github.com/streamdeck-linux-gui/streamdeck-linux-gui/releases/tag/v4.0.0).

## PIP Package

We have made some adjustments, and now everyone should be able to get the latest version when upgrading via pip. Many of you were unable to update because we had increased the version of Python to 3.11, causing pip to fetch only the latest version that was considered compatible with your Python version, even though the application works perfectly with Python 3.8.

If you initially installed via PyPI, you can now update it by running one of the following commands, depending on your installation method:

* `pip install streamdeck-linux-gui --upgrade`
* `pipx install streamdeck-linux-gui --upgrade`

## AUR Packages

We have decided to change our development workflow, and going forward, we will only use the main branch. Consequently, the `streamdeck-ui-develop` and `streamdeck-ui-beta` packages will be deprecated. From now on, please use the following packages:

* **streamdeck-ui**: Stable releases
* **streamdeck-ui-git**: Unstable, development code

We apologize for the frequent changes in AUR packages, but we are trying to determine what works best for us.

As always, if you encounter any issues, please let us know.

**Comment by xiota** [comment]
If you need my attention on something packaging related, feel free to @ me on the relevant issue.

For the deprecated packages, you can convert them into metapackages to redirect users to the most relevant package (eg, beta → stable, dev → git).  (If you like, I can provide a template.)  Then in a few weeks or months, when you've decided they're definitely not returning, you can open deletion requests.

**Reply by coolapso** [reply]
I requested the deletion of the packages and they were now deleted, think its better than cluttering AUR with unused packages.

**Comment by Danie10** [comment]
@4s3ti thanks my AUR update worked just fine. I see the Switch state that changes from 1, 2,3 etc and assume that is the Multi state change? But looking at the discussions and readme, I'm not finding anything that explains what it does and how we would use it?

**Reply by coolapso** [reply]
here: https://github.com/streamdeck-linux-gui/streamdeck-linux-gui/pull/82

maybe shoudl add them to the docs

**Comment by Danie10** [comment]
Ah thanks got it. Yes I did not look under Pull Requests. But I now see there is also a new add state + button top right above button config. So we add states there and set details for each state.

OK managed to figure out that state 0 leaves as it is, having a 2 in that field will trigger the second state, and if 2 has a state 1 then another button press reverts the button back to state 1.

So yes will be useful especially for toggle type actions like on/off, open/close, full screen/windowed, etc.

**Reply by coolapso** [reply]
exactly!! I am currently using it for switch between headphones and loudspeakers and to enable / disable casting key presses on the screen.

**Reply by coolapso** [reply]
ooh and to switch keyboard layouts xD

**Comment by oparada1988** [comment]
Great release! Thanks to all that worked on this!
